common/environment/environment.cpp

```diff
@@ -1336,13 +1538,222 @@ void CLocalEnvironment::clearCache()
+}
+
+IConstDropZoneInfoIterator * CLocalEnvironment::getDropZoneIterator() const
+{
+    return (IConstDropZoneInfoIterator *) new CConstDropZoneInfoIterator();
+}
+
+IConstMachineInfoIterator * CLocalEnvironment::getMachineIterator() const
+{
+    return (IConstMachineInfoIterator *) new CConstMachineInfoIterator();
+}
+
+//==========================================================================================
+// Iterators implementation
+
+CConstMachineInfoIterator::CConstMachineInfoIterator()
+{
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    constEnv.setown((CLocalEnvironment *)factory->openEnvironment());
+    index = 1;
+    maxIndex = constEnv->getNumberOfMachines();
+    curr = NULL;
+    if (maxIndex)
+        curr = constEnv->getMachineByIndex(index);
+}
+
+bool CConstMachineInfoIterator::first()
+{
+    return ((index == 1) && (curr != NULL));
+}
+bool CConstMachineInfoIterator::next()
+{
+    if (index < maxIndex)
+    {
+        index++;
+        curr = constEnv->getMachineByIndex(index);
+    }
+    else
+        curr = NULL;
+
+    return (curr ? true : false);
+}
+
+bool CConstMachineInfoIterator::isValid()
+{
+    return curr != NULL;
+}
+
+IConstMachineInfo & CConstMachineInfoIterator::query()
+{
+    return *curr;
+}
+
+unsigned CConstMachineInfoIterator::count()
+{
+    return maxIndex;
+}
+
+//--------------------------------------------------
+
+CConstDropZoneIteratorByComputer::CConstDropZoneIteratorByComputer(const char * computer)
+{
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    constEnv.setown((CLocalEnvironment *)factory->openEnvironment());
+    computerName.set(computer);
+    maxIndex = constEnv->getNumberOfDropZonesByComputer(computer);
+    index = 1;
+    curr = constEnv->getDropZoneByComputerByIndex(computer, index);
+}
+
+bool CConstDropZoneIteratorByComputer::first()
+{
+    return ((index == 1) && (curr != NULL));
+}
+bool CConstDropZoneIteratorByComputer::next()
+{
+    if (index < maxIndex)
+    {
+        index++;
+        curr = constEnv->getDropZoneByComputerByIndex(computerName.str(), index);
+    }
+    else
+        curr = NULL;
+
+    return (curr ? true : false);
+}
+
+bool CConstDropZoneIteratorByComputer::isValid()
+{
+    return curr != NULL;
+}
+
+IConstDropZoneInfo & CConstDropZoneIteratorByComputer::query()
+{
+    return *curr;
+}
+
+unsigned CConstDropZoneIteratorByComputer::count()
+{
+    return maxIndex;
+}
+
+IConstDropZoneInfo * CConstDropZoneIteratorByComputer::ptr()
+{
+    return curr;
+}
+
+//--------------------------------------------------
+
+CConstDropZoneInfoIterator::CConstDropZoneInfoIterator()
+{
+    Owned<IEnvironmentFactory> factory = getEnvironmentFactory();
+    constEnv.setown((CLocalEnvironment *)factory->openEnvironment());
+    maxIndex = constEnv->getNumberOfDropZones();
+    index = 1;
+    curr = constEnv->getDropZoneByIndex(index);
```

Comment:
curr should assigned in first() (so reset when first() is called).
curr should be set to NULL here.